src/libraries/System.Private.Uri/src/System/IriHelper.cs

```diff
@@ -226,56 +218,61 @@ internal static unsafe string EscapeUnescapeIri(char* pInput, int start, int end
                 {
                     // unicode
 
-                    char ch2;
+                    bool escape;
+                    bool surrogatePair = false;
+
+                    char ch2 = '\0';
 
                     if ((char.IsHighSurrogate(ch)) && (next + 1 < end))
                     {
                         ch2 = pInput[next + 1];
                         escape = !CheckIriUnicodeRange(ch, ch2, ref surrogatePair, component == UriComponents.Query);
-                        if (!escape)
-                        {
-                            // copy the two chars
-                            dest.Append(pInput[next++]);
-                            dest.Append(pInput[next]);
-                        }
                     }
                     else
                     {
-                        if (CheckIriUnicodeRange(ch, component == UriComponents.Query))
+                        escape = !CheckIriUnicodeRange(ch, component == UriComponents.Query);
+                    }
+
+                    if (escape)
+                    {
+                        Span<byte> encodedBytes = stackalloc byte[4];
+
+                        Rune rune;
+                        if (surrogatePair)
                         {
-                            // copy it
-                            dest.Append(pInput[next]);
+                            rune = new Rune(ch, ch2);
                         }
-                        else
+                        else if (!Rune.TryCreate(ch, out rune))
                         {
-                            // escape it
-                            escape = true;
+                            rune = Rune.ReplacementChar;
```

Comment:
Is this new behavior?

Do we understand under what circumstances `Rune.TryCreate` might fail?

Reply:
> Is this new behavior?

No, Utf8Encoding will use the replacement char as well if it fails to encode a character. Specifically, it will use it if you try to encode a single character that is also a high/low surrogate.